### PR TITLE
fix(artifactType): resolve mixing between artifact types and version

### PR DESCRIPTION
**Pull request by gal-yardeni**
Recently we came across a delivery config that had both docker and Debian artifacts, and keel didn't manage it correctly.
A log message helped us discovered some mixup between artifact types and versions when queuing versions for approval. 

 -- for example:
docker artifact with version 1.0
Debian artifact with version 2.0
both were approved for `testing` environment
our current flow will fetch all `queuedForApproval` versions and returned both 1.0 and 2.0, and then will `checkStatelessConstraints` with each artifact type (like docker with version 2.0), which will potentially be an error.

The problem was that we don't take into consideration a unique artifact identifier like `reference` when fetching all `queuedForApproval` versions.
This fix adds a new column for artifact reference to the `environment_artifact_queued_approval` table, which will enforce considering the right artifact per version.

[Update] This PR is also addressing issue #1270 by adding a `reference` column to `environment_artifact_constraint` table as well, and ensuring a unique and correct fetch by artifact version + artifact reference combination. 

[Update #2] at first I was using `artifact_type` as an identifier, but after a discussion with the team we've realized a single config can have multiple artifact types with the same version, BUT their reference cannot be the same. So I converted `type` into `reference`.

### CODE REVIEW

File: keel-core/src/test/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentPromotionCheckerTests.kt [hunk 1]
```diff
@@ -333,5 +351,102 @@ internal class NewEnvironmentPromotionCheckerTests : JUnit5Minutests {
...
+      }
+
+      test("verify the right artifact type and version approved") {
+        verify {
+          repository.approveVersionFor(deliveryConfigWith2ArtifactTypes, dockerArtifact, "2.0", environment.name)
+          repository.approveVersionFor(deliveryConfigWith2ArtifactTypes, debianArtifact, "3.0", environment.name)
+        }
+      }
+
+      test("approved versions of other artifact types are not getting mixed up and are not checked") {
```
Comment (gal-yardeni) [inline keel-core/src/test/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentPromotionCheckerTests.kt:409]: Prior to this change, this test failed as different artifact types and versions got checked.

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlDeliveryConfigRepository.kt [hunk 2]
```diff
@@ -529,6 +530,7 @@ class SqlDeliveryConfigRepository(
                 .set(ENVIRONMENT_ARTIFACT_QUEUED_APPROVAL.ENVIRONMENT_UID, envUid)
```
Comment (emjburns) [inline keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlDeliveryConfigRepository.kt:530]: you can delete the comment on line 528! :)

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlDeliveryConfigRepository.kt [hunk 3]
```diff
@@ -897,14 +917,15 @@ class SqlDeliveryConfigRepository(
   }
 
   // todo eb: add link to artifact https://github.com/spinnaker/keel/issues/1270
```
Comment (emjburns) [inline keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlDeliveryConfigRepository.kt:919]: You can delete this comment now! :)

File: keel-sql/src/main/resources/db/changelog/20200730-add-artifact-type-column.yml [hunk 4]
```diff
@@ -0,0 +1,25 @@
+databaseChangeLog:
+  - changeSet:
+      id: add-artifact-type-column
+      author: lpollo
```
Comment (emjburns) [inline keel-sql/src/main/resources/db/changelog/20200730-add-artifact-type-column.yml:4]: this should be you! not that it really matters

File: keel-sql/src/main/resources/db/changelog/20200730-add-artifact-type-column.yml [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+      changes:
+        - addColumn:
+            tableName: environment_artifact_queued_approval
+            columns:
+              - name: artifact_type
+                type: varchar(255)
```
Comment (emjburns) [inline keel-sql/src/main/resources/db/changelog/20200730-add-artifact-type-column.yml:10]: I think this should have:
`afterColumn: artifact_version` so it's not at the end of the table.

Also, do you ever query using artifact type? we might need an index on it if you do anything like "get all pending debians for this environment"

File: keel-sql/src/main/resources/db/changelog/20200730-add-artifact-type-column.yml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+                constraints:
+                  nullable: false
+        - addColumn:
+            tableName: environment_artifact_constraint
+            columns:
+              - name: artifact_type
+                type: varchar(255)
```
Comment (emjburns) [inline keel-sql/src/main/resources/db/changelog/20200730-add-artifact-type-column.yml:17]: same `afterColumn` comment + question about indexing here as well

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlDeliveryConfigRepository.kt [hunk 7]
```diff
@@ -897,14 +917,15 @@ class SqlDeliveryConfigRepository(
...
-  override fun getQueuedConstraintApprovals(deliveryConfigName: String, environmentName: String): Set<String> {
+  override fun getQueuedConstraintApprovals(deliveryConfigName: String, environmentName: String, artifactType: String): Set<String> {
     val environmentUID = environmentUidByName(deliveryConfigName, environmentName)
       ?: return emptySet()
 
     return sqlRetry.withRetry(READ) {
       jooq.select(ENVIRONMENT_ARTIFACT_QUEUED_APPROVAL.ARTIFACT_VERSION)
         .from(ENVIRONMENT_ARTIFACT_QUEUED_APPROVAL)
         .where(ENVIRONMENT_ARTIFACT_QUEUED_APPROVAL.ENVIRONMENT_UID.eq(environmentUID))
+        .and(ENVIRONMENT_ARTIFACT_QUEUED_APPROVAL.ARTIFACT_TYPE.eq(artifactType))
```
Comment (emjburns) [inline keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlDeliveryConfigRepository.kt:928]: adding this here means that when you roll this out we won't consider older versions with `artifactType == null` (all the old data). It's probably worth allowing null here as well

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlDeliveryConfigRepository.kt [hunk 8]
```diff
@@ -937,15 +960,17 @@ class SqlDeliveryConfigRepository(
...
   ) {
     sqlRetry.withRetry(WRITE) {
       environmentUidByName(deliveryConfigName, environmentName)
         ?.also { envUid ->
           jooq.deleteFrom(ENVIRONMENT_ARTIFACT_QUEUED_APPROVAL)
             .where(
               ENVIRONMENT_ARTIFACT_QUEUED_APPROVAL.ENVIRONMENT_UID.eq(envUid),
-              ENVIRONMENT_ARTIFACT_QUEUED_APPROVAL.ARTIFACT_VERSION.eq(artifactVersion))
+              ENVIRONMENT_ARTIFACT_QUEUED_APPROVAL.ARTIFACT_VERSION.eq(artifactVersion),
+              ENVIRONMENT_ARTIFACT_QUEUED_APPROVAL.ARTIFACT_TYPE.eq(artifactType))
```
Comment (emjburns) [inline keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlDeliveryConfigRepository.kt:973]: I think you'll also have to allow `null` here to be able to clean up old data

File: keel-sql/src/main/resources/db/changelog/20200730-add-artifact-type-column.yml [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+                constraints:
+                  nullable: false
+
+    rollback:
```
Comment (emjburns) [inline keel-sql/src/main/resources/db/changelog/20200730-add-artifact-type-column.yml:21]: You can do:
```      rollback:
        - dropColumn:
            tableName: environment_artifact_constraint
            columnName: artifact_type```
for both, you don't have to write the raw sql

**Comment by emjburns** [comment]
Some small comments for you. My main question is - have you made sure this works in all forms on old data? I don't want there to be a problem rolling this out because we fail when `artifact_type` is `null`.

File: keel-sql/src/main/resources/db/changelog/20200730-add-artifact-type-column.yml (hunk 5 above)
Comment (gal-yardeni) [reply keel-sql/src/main/resources/db/changelog/20200730-add-artifact-type-column.yml:10]: oh i didn't know about `afterColumn` trick :) 
I'm not querying by type only, we query by type + version + artifact type+ environment uid from what I can tell.

Comment (robfletcher) [reply keel-sql/src/main/resources/db/changelog/20200730-add-artifact-type-column.yml:10]: Sounds like it should be added into that index then, no?

Comment (gal-yardeni) [reply keel-sql/src/main/resources/db/changelog/20200730-add-artifact-type-column.yml]: added.

**Comment by gal-yardeni** [comment]
@emjburns @robfletcher I've changed `type` into `reference` as we discussed on slack. I added the index and made `artifact_referance` nullable for backwards compatibility. Let me know if anything else is missing! thanks

File: keel-sql/src/main/resources/db/changelog/20200731-add-artifact-reference-column.yml [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+                  nullable: true
+      rollback:
+        - dropColumn:
+            tableName: environment_artifact_constraint
+            columnName: artifact_reference
+        - dropColumn:
+            tableName: environment_artifact_queued_approval
+            columnName: artifact_reference
+
+  - changeSet:
```
Comment (gal-yardeni) [inline keel-sql/src/main/resources/db/changelog/20200731-add-artifact-reference-column.yml:30]: for some reason, the second part `create-artifact-reference-indicies` is not running. What did I miss?

File: keel-sql/src/main/resources/db/changelog/20200731-add-artifact-reference-column.yml [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+                afterColumn: artifact_version
+                constraints:
+                  nullable: true
+      rollback:
+        - dropColumn:
+            tableName: environment_artifact_constraint
+            columnName: artifact_reference
+        - dropColumn:
+            tableName: environment_artifact_queued_approval
+            columnName: artifact_reference
```
Comment (robfletcher) [inline keel-sql/src/main/resources/db/changelog/20200731-add-artifact-reference-column.yml:28]: I don't think you need this. Most Liquibase migrations have auto-rollback. It's only really when you're doing something unusual (e.g. with `sql`) that you need an explicit rollback.

File: keel-core/src/test/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentConstraintRunnerTests.kt [hunk 12]
```diff
@@ -68,6 +68,7 @@ internal class EnvironmentConstraintRunnerTests : JUnit5Minutests {
       deliveryConfig.name,
       environment.name,
       "2.0",
+      artifact.type,
```
Comment (robfletcher) [inline keel-core/src/test/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentConstraintRunnerTests.kt:71]:
```suggestion
      artifact.reference,
```

File: keel-core/src/test/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentConstraintRunnerTests.kt [hunk 13]
```diff
@@ -76,6 +77,7 @@ internal class EnvironmentConstraintRunnerTests : JUnit5Minutests {
       deliveryConfig.name,
       environment.name,
       "1.2",
+      artifact.type,
```
Comment (robfletcher) [inline keel-core/src/test/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentConstraintRunnerTests.kt:80]:
```suggestion
      artifact.reference,
```

File: keel-core/src/test/kotlin/com/netflix/spinnaker/keel/constraints/ManualJudgementConstraintEvaluatorTests.kt [hunk 14]
```diff
@@ -50,6 +50,7 @@ internal class ManualJudgementConstraintEvaluatorTests : JUnit5Minutests {
       deliveryConfigName = configName,
       environmentName = env.name,
       artifactVersion = version,
+      artifactReference = artifact.type,
```
Comment (robfletcher) [inline keel-core/src/test/kotlin/com/netflix/spinnaker/keel/constraints/ManualJudgementConstraintEvaluatorTests.kt:53]:
```suggestion
      artifactReference = artifact.reference,
```

File: keel-sql/src/main/resources/db/changelog/20200731-add-artifact-reference-column.yml (hunk 11 above)
Comment (gal-yardeni) [reply keel-sql/src/main/resources/db/changelog/20200731-add-artifact-reference-column.yml:28]: thanks! is this true for the index rollback below as well?

**Pull request merged**